src/mcts/search.cc

```diff
@@ -1048,285 +1151,624 @@ void SearchWorker::GatherMinibatch() {
+      picked_node.ooo_completed = true;
     }
-    // Check for stop at the end so we have at least one node.
-    if (search_->stop_.load(std::memory_order_acquire)) return;
   }
 }
 
-namespace {
-void IncrementNInFlight(Node* node, Node* root, int amount) {
-  if (amount == 0) return;
-  while (true) {
-    node->IncrementNInFlight(amount);
-    if (node == root) break;
-    node = node->GetParent();
+void SearchWorker::PickNodesToExtend(int collision_limit) {
+  {
+    task_count_.store(0, std::memory_order_release);
+    next_task_available_.store(0, std::memory_order_release);
+    completed_tasks_.store(0, std::memory_order_release);
+    picking_tasks_.clear();
+    // Reserve because resizing breaks pointers held by the task threads.
+    picking_tasks_.reserve(100);
+    // While nothing is ready yet - wake the task runners so they are ready to
+    // receive quickly.
+    std::unique_lock<std::mutex> lock(picking_tasks_mutex_);
+    task_added_.notify_all();
+  }
+  std::vector<Move> empty_movelist;
+  // This lock must be held until after the task_completed_ wait succeeds below.
+  // Since the tasks perform work which assumes they have the lock, even though
+  // actually this thread does.
+  SharedMutex::Lock lock(search_->nodes_mutex_);
+  PickNodesToExtendTask(search_->root_node_, 0, collision_limit, empty_movelist,
+                        &minibatch_, &main_workspace_);
+  {
+    // Spin lock, other tasks should be done soon.
+    while (true) {
+      int completed = completed_tasks_.load(std::memory_order_acquire);
+      int todo = task_count_.load(std::memory_order_acquire);
+      if (todo == completed) break;
+      SpinloopPause();
+    }
+    for (int i = 0; i < picking_tasks_.size(); i++) {
+      for (int j = 0; j < picking_tasks_[i].results.size(); j++) {
+        minibatch_.emplace_back(std::move(picking_tasks_[i].results[j]));
+      }
+    }
   }
 }
-}  // namespace
 
-// Returns node and whether there's been a search collision on the node.
-SearchWorker::NodeToProcess SearchWorker::PickNodeToExtend(
-    int collision_limit) {
-  // Starting from search_->root_node_, generate a playout, choosing a
-  // node at each level according to the MCTS formula. n_in_flight_ is
-  // incremented for each node in the playout (via TryStartScoreUpdate()).
+void SearchWorker::PickNodesToExtendTask(Node* node, int base_depth,
+                                         int collision_limit,
+                                         const std::vector<Move>& moves_to_base,
+                                         std::vector<NodeToProcess>* receiver,
+                                         TaskWorkspace* workspace) {
+  // TODO: Bring back pre-cached nodes created outside locks in a way that works
+  // with tasks.
+  // TODO: pre-reserve visits_to_perform for expected depth and likely maximum
+  // width. Maybe even do so outside of lock scope.
+  std::vector<std::unique_ptr<int[]>> visits_to_perform;
```

Comment:
I started with std::vector<std::vector<int>> - I switched to this when I decided to make them all fixed length and track the populated amount externally to vector in order to avoid paying clearing costs when I could just reset the tracked length.  I'll try std::array in 1490.